sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -939,68 +958,73 @@ class GpuDynamicPartitionDataConcurrentWriter(
+        withResource(GpuColumnVector.from(batch)) { table =>
+          val splits =
+            splitToFitMaxRecords(
+              table,
               maxRecordsPerFile,
-              status.writerStatus.recordsInFile,
-              table.getRowCount()
-            )
-
-            val dataTypes = GpuColumnVector.extractTypes(batch)
-
-            var needNewWriter = status.writerStatus.recordsInFile >= maxRecordsPerFile
-            withResource(table.contiguousSplit(splitIndexes: _*)) {tabs =>
-              tabs.foreach(b => {
-                if (needNewWriter) {
-                  status.writerStatus.fileCounter += 1
-                  assert(status.writerStatus.fileCounter <= MAX_FILE_COUNTER,
-                    s"File counter ${status.writerStatus.fileCounter} " +
-                      s"is beyond max value $MAX_FILE_COUNTER")
-                  status.writerStatus.outputWriter.close()
-
-                  // start a new writer
-                  val w = newWriter(partitionDir, None, status.writerStatus.fileCounter)
-                  status.writerStatus.outputWriter = w
-                  status.writerStatus.recordsInFile = 0L
-                }
-                val cb = withResource(b.getTable()) {tab =>
-                  GpuColumnVector.from(tab, dataTypes)
-                }
-                statsTrackers.foreach(_.newBatch(status.writerStatus.outputWriter.path(), cb))
-                status.writerStatus.recordsInFile += b.getRowCount()
-                status.writerStatus.outputWriter.writeAndClose(cb, statsTrackers)
-                needNewWriter = true
-              })
-            }
+              status.writerStatus.recordsInFile)
+          (splits, GpuColumnVector.extractTypes(batch))
+        }
+      }
+      var needNewWriter = status.writerStatus.recordsInFile >= maxRecordsPerFile
+      withResource(tabs) { _ =>
+        tabs.zipWithIndex.foreach { case (b, partIndex) =>
+          if (needNewWriter) {
+            status.writerStatus.fileCounter += 1
+            assert(status.writerStatus.fileCounter <= MAX_FILE_COUNTER,
+              s"File counter ${status.writerStatus.fileCounter} " +
+                  s"is beyond max value $MAX_FILE_COUNTER")
+            status.writerStatus.outputWriter.close()
+
+            // start a new writer
+            val w = newWriter(partitionDir, None, status.writerStatus.fileCounter)
+            status.writerStatus.outputWriter = w
+            status.writerStatus.recordsInFile = 0L
           }
+          val cb = GpuColumnVector.from(b.getTable, dataTypes)
+          closeOnExcept(cb) { _ =>
+            statsTrackers.foreach(_.newBatch(status.writerStatus.outputWriter.path(), cb))
+            status.writerStatus.recordsInFile += b.getRowCount
+          }
+          // close the contiguous table
+          tabs(partIndex) = null
+          withResource(b) { _ =>
+            status.writerStatus.outputWriter.writeAndClose(cb, statsTrackers)
+          }
```

Comment:
Holding `b` open but not referencing it in the block? Seems similar to other comment, need to close as soon as `cb` is built.

Reply:
yes, I got confused with ownership here: we took out the columns from the Table, and at that point it doesn't matter that we closed the ContiguousTable or the table_view.